apps/apps.c

```diff
@@ -2343,6 +2343,69 @@ int app_isdir(const char *name)
 }
 #endif
 
+/* app_dirname section */
+/*
+ * TODO: implement app_dirname for Windows
+ * and VMS.
+ */
+#if !defined(_WIN32) && !defined(__VMS)
+/*
+ * This exactly follows what POSIX's
+ * dirname does, but is implemented
+ * in a more platform independent way.
+ *
+ * path        dirname
+ * /usr/lib    /usr
+ * /usr/       /
+ * usr         .
+ * /           /
+ * .           .
+ * ..          .
+ * NULL        .
+ * ""          .
+ *
+ * Note: this function also keeps the
+ * possibility of modifying the 'path'
+ * string same as POSIX dirname.
+ */
+char *app_dirname(char *path)
```

Comment:
Actually, you could rename this to `fallback_dirname` with no guards, make i static, and then have this code:

``` C
#if !defined(_WIN32) && !defined(__VMS)
char *app_dirname(char *path)
{
    return fallback_dirname(path);
}
#endif
```

At least in the VMS case, I would find the fallback useful, in case someone uses Unix syntax in their file specs (it's actually legal, and `openssl` does understand such file specs)